l3backend/l3backend-color.dtx

```diff
@@ -1248,6 +1253,401 @@
+% underlying idea comes from Heiko Oberdiek's \pkg{luacolor} package.
+%
+% \begin{macro}{ltx.color}
+% \begin{macro}{ltxcolor}
+%   The basic table should be set up in \pkg{l3luatex}, but we want a separate
+%   sub-table for color.
+%    \begin{macrocode}
+ltx = ltx or {color = {}}
+ltx.color = ltx.color or {}
+local ltxcolor = ltx.color
+%    \end{macrocode}
+% \end{macro}
+% \end{macro}
+%
+% Load basic support and print some information.
+%    \begin{macrocode}
+require("ltluatex")
+local luatexbase = luatexbase
+luatexbase.provides_module({name = "l3backend-luatex", date = "2021-06-03"})
+%    \end{macrocode}
+%
+% \begin{macro}{attr}
+%   An attribute to track color. There is no need to access this from \TeX{} so
+%   in contrast to \pkg{luacolor} we just allocate directly in Lua.
+%    \begin{macrocode}
+local color_attr = luatexbase.new_attribute("color")
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{map}
+%   A map between color strings and their color_index, plus the tracking number.
+%    \begin{macrocode}
+local map = {n = 0}
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{color_index}
+%   Get the color_index of a color or create a new color_index if necessary.
+%    \begin{macrocode}
+local function color_index(color)
+  local n = map[color]
+  if not n then
+    n = map.n + 1
+    map[n] = color
+    map[color] = n
+  end
+  return n
+end
+%    \end{macrocode}
+% \end{macro}
+%
+%   Local copies of global tables.
+%    \begin{macrocode}
+local node   = node
+local string = string
+local tex    = tex
+local texio  = texio
+%    \end{macrocode}
+%
+%   Local copies of standard functions.
+%    \begin{macrocode}
+local getbox           = tex.getbox
+local has_attribute    = node.has_attribute
+local id               = node.id
+local insert_before    = node.insert_before
+local module_error     = luatexbase.module_error
+local node_new         = node.new
+local node_traverse    = node.traverse
+local saveboxresources = tex.saveboxresources
+local set_attribute    = tex.setattribute
+local type             = node.type
+local subtype          = node.subtype
+%    \end{macrocode}
+%
+% \begin{macro}{ltx.color.select}
+%   Get the color_index of a color or create a new color_index if necessary.
+%    \begin{macrocode}
+local function select(color_str)
+  set_attribute(color_attr,color_index(color_str))
+end
+ltxcolor.select = select
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{node_map, node_types,literals}
+%   We abstract the idea of different types of node where it makes sense:
+%   notice for example that in contrast to \pkg{luacolor} we extract some
+%   one-off information in the function that needs it. We also use a single
+%   local \enquote{map} for the node values.
+%    \begin{macrocode}
+local node_map = {
+    list         = 1 ,
+    list_leaders = 2 ,
+    list_disc    = 3 ,
+    color        = 4 ,
+    no_color     = 5
+  }
```

Comment:
IMO these should just be separate variables since we never use `node_map` except for accessing an explicitly specified elements.

Reply:
They can be - that's of course what Heiko does. I felt that as this was one 'idea' (mapping from node type to number), it should be in one place. It's not hard to alter - something like `node...` so they have a common pattern?

Reply:
>  felt that as this was one 'idea' (mapping from node type to number), it should be in one place.

Then let's leave it like this. It's probably mostly a style question anyway, any performance impact is probably non-measurable anyway.

Reply:
I'm happy to alter: I'd settled on `id_<name>` ...